### ISSUE TITLE
softmax dimension seems not right

### ISSUE THREAD

**Issue by Primus-zhao** [timomernick/pytorch-capsule#7]
Notice that in capsule_layer.py, c_ij is generated by following eq:

c_ij = F.softmax(b_ij)

which softmax the vector based on dim 1, which is input channel. The result is softmax b_ij in "i" dimension, but i think we should softmax in "j" dimension according to original paper, so i think it should be as following?(minimal change)

c_ij = F.softmax(b_ij.transpose(1, 2)).transpose(1, 2)

**Comment by Primus-zhao** [comment]
Sorry, my mistake, you code is right.

**Comment by XiongweiWu** [comment]
@Primus-zhao I think your idea is correct, why later u mean it's your mistake?

**Comment by Primus-zhao** [comment]
@XiongweiWu Nothing important, in this implementation, if we use correct softmax dimension, the convergence speed becomres unbearably slow.  I think it's due to weight initialization problem and ignore of disabling backpro in initial dynamic routing process. 
Although in the original paper Hinton use reverse softmax, it's highly possible suffering from saturation problem in initial training stage unless we scales net weights carefully. In fact, i've tested various schemes and found that their results are quite close to each other, so i think it's wise to keep this scheme rather than changes bunch of related parts.
However, if you insist on using the original implement, you can check the tensorflow version written by naturomics or pytorch version under my git branch, which follows the original paper as well as gives same results with acceptable convergence speed.

**Comment by XiongweiWu** [comment]
@Primus-zhao According to my experiments, the convergence speed is almost the same for these two settings, both on cifar10 and mnist. One key shortcoming of the original softmax setting is if we do not fix the digit coefficients(mean b_ij), the training process will become extremely unstable(<1e-4 lr can converge) and I assume it's larger norm of b_ij that causes this problem. I just curious about why the other setting can also work very well since for every target class the total weights of a single image feature map is the same. But intuitively it's also reasonable, because some zero-norm capsules may receive high weight for incorrect target classes.

**Comment by sairaamVenkatraman** [comment]
@XiongweiWu By fixing the mean b_ij, do you mean forcing it to zero? Also, @Primus-zhao  what scaling do you do on the net weights?

**Comment by Primus-zhao** [comment]
@sairaamVenkatraman ,it's been a long time so i'm not sure i can give the accurate answer, i guess it's 0.1
